### DISCUSSION TITLE
Index major versions of clojure.core

### DISCUSSION THREAD

**Discussion by raymcdermott** [repl-acement/repl-acement#8] [Ideas]
To provide the core functions all have IDs and digests per change

**Comment by andrewzhurov** [comment]
If I get it right, indexing of major clojure versions would be required if we are to reference functions by uuids, so codebases can agree on what uuid=fn.
However, could each codebase idempotently derive an id of a function out of function's body?
I.e., can it be that `(id (fn [ ]))`, wherever it's ran, results in the same value?
Having that would mean that any codebase can index (derive an id of) any function (be it a core one or from a lib) ad-hoc, as they use it, without the need to coordinate with other folks, as they can derive the same id for that function.

**Reply by yenda** [reply]
The function you are describing is a digest. The digest changes when something changes, the ID needs to be stable, if you use a digest for ID, you can't tell what is what after a refactoring. That's why you need both, an ID that identifies your thing and a digest that tells you when it changes.

**Reply by andrewzhurov** [reply]
> the ID needs to be stable, if you use a digest for ID, you can't tell what is what after a refactoring

Hmm.. why would we want to know that?

**Reply by raymcdermott** [reply]
Keeping a history from 0 to now is interesting as the code evolves. Git already does this per ns. We want to do that per var since we have semantic knowledge that GitHub does not. Have you seen the [codeq project blog](https://blog.datomic.com/2012/10/codeq.html)? Rich explains this aspect over there.

**Reply by raymcdermott** [reply]
Maybe another question is whether we import the entire history and mark the releases by tags or just import the data based on the tags and ignore intervening commits? I'm gonna to say the latter has the most utility.

**Reply by yenda** [reply]
> Maybe another question is whether we import the entire history and mark the releases by tags or just import the data based on the tags and ignore intervening commits? I'm gonna to say the latter has the most utility.


We probably want to leave that choice to the importer, project workflow varies a lot from project to project. There's definitely value in skipping for project that tend to commit non-functional code to master as part of a series of commits, but I also see value in importing every commit, especially once the project is imported and we are listening on new commits.

**Reply by andrewzhurov** [reply]
> Keeping a history from 0 to now is interesting as the code evolves. Git already does this per ns. We want to do that per var since we have semantic knowledge that GitHub does not. Have you seen the [codeq project blog](https://blog.datomic.com/2012/10/codeq.html)? Rich explains this aspect over there.

Thanks for the link
I guess there are gains from keeping history, for things akin to:
- git log
- git bisect
- knowing that a dependent library has a new version / there are functions that are (possibly) better than the ones you use

That are known to be useful.

I'm not certain that there needs to be an UUID for that. It would seem that for the forementioned features datamodel of `<fn2>` superseeds `<fn1>` would be sufficient

What are benefits of having such a datamodel?
```
<fn1> is <uuid1>
<fn2> is <uuid1>
<fn2> superseeds <fn1>
```

Perhaps a downside would be that `<uuid1>` needs to in consensus, so we'd have an issue if user1 and user2 wants to publish simultaneously (e.g., user1 indexes clj-http v1 and user2 indexes clj-http v2, giving clj-http.core/post `<uuid1>` and `<uuid2>` correspondingly, whereas it needed to be of the same value, e.g., `<uuid1>` in both cases).
Although, I guess, uuid can be = `(hash fn-name)`. E.g., `(hash 'clj-http.core/post)`. Or it can be the name itself. So it's not much of a problem, I guess. The question why do we need it stands though

**Reply by andrewzhurov** [reply]
> Maybe another question is whether we import the entire history and mark the releases by tags or just import the data based on the tags and ignore intervening commits? I'm gonna to say the latter has the most utility.

This is an interesting question
I would be happy to discuss it, would we like to have it in a separate thread? Keeping this one focused on `uuid`, problem it solves and possible alternative solutions

**Reply by andrewzhurov** [reply]
* library is a dictionary from names to values
  * e.g.,
  ```
  {:library/name 'org.clojure
   :library/version 11
   :library/names {'clojure.core/+ <hash>}}
  ```
  * there can be a separate history entity, akin to git commits data model, to track how values under names evolve
    ```
    {:commit/library <library of org.clojure 11>
     :commit/parent <commit of org.clojure 10>}
    ```
  * there can be a mutable name that points to the latest history element
    * e.g., https://clojure.org/libraries/org.clojure/latest
  * how to name things should be up to the end-user
    * so folks can have their personal dictionaries
      * e.g., after loading Clojure 11 dictionary and refering to `<hash of clojure.core/+>` in their codebase, they could switch to their personal dictionary of
        ```
        {:library/name 'local
         :library/names {'sum <hash of clojure.core/+>}}
        ```
        And now the same function can be referred to by two names, but will be displayed as `sum`, since that is user's preference.
      * akin to how it's possible in Unisonlang
    * in order to get value of clojure.core/+ one could get a dictionary, e.g., of clojure 1.11, and lookup the value under the name
      * how a user could be sure that a dictionary they're using came from an authority?
        * e.g., how a user can be sure that a dictionary of Clojure 11 came from Clojure authorities?
        * so there can't be an attack when one puts their malicious values under names
          * e.g., `clojure.core/+` pointing to `(fn some-malicious-+ ...)`
          * and others use it believing it came from authorities
        * have security on locations
          * i.e., stuff that comes from that location is legit
          * e.g., stuff that comes from https://clojure.org/libraries/org.clojure/11 is legit
          * a person would need to fetch stuff from that location in order to be sure that it's legit
            * this requires location to be always accessible
        * have security on data
          * i.e., have authorities sign below libraries
          * e.g., having Clojure publish
            ```
            {:authorization/content <clojure-11-library>
             :authorization/signature (sign <clojure-11-library>)
             :authorization/authority <clojure>}
            ```
          * then whoever has content and signature can verify that it's been authorized by the authority
            * this allows peers share libraries between themselves and be sure they are legit
